### PR TITLE
Elasticsearch addon

### PR DESCRIPTION
**Pull request by larmog**
Initial PR for collaboration.

### CODE REVIEW

File: addons/elasticsearch.yml [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: elasticsearch
+  labels:
+    component: elasticsearch
+    role: client
+spec:
+  type: LoadBalancer
```
Comment (luxas) [inline addons/elasticsearch.yml]: Is this directly copied from somewhere? I don't think `LoadBalancer` exists on bare-metal

File: addons/elasticsearch.yml [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    role: client
+  ports:
+  - name: http
+    port: 9200
+    protocol: TCP
+---
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: es-master
```
Comment (luxas) [inline addons/elasticsearch.yml]: Why not in `kube-systemd`?

File: addons/elasticsearch.yml [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    spec:
+      serviceAccount: elasticsearch
+      containers:
+      - name: es-master
+        securityContext:
+          privileged: true
+          capabilities:
+            add:
+              - IPC_LOCK
+        image: larmog/elasticsearch-armhf:2.2.1
```
Comment (luxas) [inline addons/elasticsearch.yml]: `kubernetesonarm/elasticsearch` and `imagePullPolicy: IfNotPresent`

File: addons/elasticsearch.yml [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        env:
+        - name: KUBERNETES_CA_CERTIFICATE_FILE
+          value: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: "CLUSTER_NAME"
```
Comment (luxas) [inline addons/elasticsearch.yml]: is the quoting required?

File: addons/elasticsearch.yml [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        ports:
+        - containerPort: 9300
+          name: transport
+          protocol: TCP
+        volumeMounts:
+        - mountPath: /data
+          name: storage
+      volumes:
+          - emptyDir:
+              medium: ""
```
Comment (luxas) [inline addons/elasticsearch.yml]: What does this line do?

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,33 @@
+FROM larmog/armhf-alpine-java:jdk-8u73
+
+EXPOSE 9200 9300
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: Not required, makes it only slower :)

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+ENV VERSION 2.2.1
+
+# Install Elasticsearch.
+RUN apk add --update curl ca-certificates sudo && \
+  ( curl -Lskj https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/$VERSION/elasticsearch-$VERSION.tar.gz | \
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: `curl -sSL https://download.elasticsearch.org/elasticsearch/release/org/elasticsearch/distribution/tar/elasticsearch/$VERSION/elasticsearch-$VERSION.tar.gz | tar -xz --strip-components=1 -C es` makes the code much easier

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  gunzip -c - | tar xf - ) && \
+  mv /elasticsearch-$VERSION /elasticsearch && \
+  rm -rf $(find /elasticsearch | egrep "(\.(exe|bat)$|sigar/.*(dll|winnt|x86-linux|solaris|ia64|freebsd|macosx))") && \
+  apk del curl
+
+# Volume for Elasticsearch data
+VOLUME ["/data"]
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: not needed

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+# Copy configuration
+ADD config/logging.yml /elasticsearch/config
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: I think `COPY` is preferred over `ADD`

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,33 @@
+FROM larmog/armhf-alpine-java:jdk-8u73
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: I'm already working on a `luxas/java` image for this, will upload when I've got `libc` compiled for alpine.

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 6 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: You mean java? :)

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 10 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: I know and we can change it when your done. But for now ...

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 6 above)
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: No `EXPOSE` adds an extra layer that's only required for `docker run -P`. The same for volume. I'm trying to make the layers as few as possible.

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 10 above)
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: Yep, fine for now.

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 6 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: The 9200 port is the default es port and 9300 is for node discovery in a es cluster. You need them.

Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: Yes, but they don't have to be defined in the Dockerfile for docker or Kubernetes to open the ports. Port 9200 and 9300 will be exposed without this layer.

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 7 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: Yes agree, much of this comes from (@)pires. I've just rearranged stuff.

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 8 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: What you mean is not needed?

File: addons/elasticsearch.yml (hunk 5 above)
Comment (larmog) [inline addons/elasticsearch.yml]: You mean `medium: ""`?

File: images/kubernetesonarm/elasticsearch/Dockerfile (hunk 6 above)
Comment (larmog) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: But it tells your users of your image that it expects these ports to be exposed. If you are looking for optimization then perhaps a better solution is to flatten you image after build.

**Comment by larmog** [comment]
Fixed, Ok?

**Comment by luxas** [comment]
I'm on a journey until monday, will take a look after that

File: addons/elasticsearch.yaml [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+          value: false
+        ports:
+        - containerPort: 9300
+          name: transport
+          protocol: TCP
+        volumeMounts:
+        - mountPath: /data
+          name: storage
+      volumes:
+          - emptyDir:
```
Comment (luxas) [inline addons/elasticsearch.yaml]: `emptyDir: {}` for consistency

File: addons/elasticsearch.yaml [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    spec:
+      serviceAccount: elasticsearch
+      containers:
+      - name: es-master
+        securityContext:
+          privileged: true
+          capabilities:
+            add:
+              - IPC_LOCK
+        image: kubernetesonarm/elasticsearch:2.2.1
```
Comment (luxas) [inline addons/elasticsearch.yaml]: Please `kubernetesonarm/elasticsearch` only and `imagePullPolicy: IfNotPresent` like other manifests

File: images/kubernetesonarm/elasticsearch/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+ENV VERSION 2.2.1
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/Dockerfile]: Please add `ELASTICSEARCH_VERSION` to `images/version.sh`, `COPY version.sh /`
Copy it to current dir via a `build.sh` file first.
Then `RUN . /version.sh && ... `
The `go` image is set up this way, you may find inspiration there (or I may do it as a followup)

File: images/kubernetesonarm/elasticsearch/run.sh [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+# Kubernetes stuff
+export NAMESPACE=${NAMESPACE:-default}
+export DISCOVERY_SERVICE=${DISCOVERY_SERVICE:-elasticsearch-discovery}
+
+# allow for memlock
+ulimit -l unlimited
+
+# run
+sudo -E -u elasticsearch /elasticsearch/bin/elasticsearch
```
Comment (luxas) [inline images/kubernetesonarm/elasticsearch/run.sh]: I'm not an expert at this, but why is `sudo` required?
I see you're adding an es user, but why not use it directly as `root`?

**Comment by luxas** [comment]
LGTM overall, some small things only :)

**Comment by luxas** [comment]
LGTM

I'll fix the comments in a follow-up commit.

**Comment by larmog** [comment]
Sorry, I wasn't finished :) I've been having some issues with my cluster and had to spend some time to fix it. I had some issues with `elasticsearch`, I couldn't get it stable. I haven't been able to get it to run stable for a longer period. Now I've added three more `odroid-c2` boards, so let's see if it makes any difference :). `Logstash-forwarder` wasn't a hit, it practically occupied the whole node. I've switched to `fluentd` and it looks more promising, but I can't say for sure.

**Comment by luxas** [comment]
Okay, send follow up PRs when you're ready. It's on `dev`, so it's not such a big deal if it's unstable

**Pull request merged**